process/sync/storageBootstrap/shardStorageBootstrapper.go

```diff
@@ -139,6 +140,80 @@ func (ssb *shardStorageBootstrapper) cleanupNotarizedStorage(shardHeaderHash []b
 	}
 }
 
+func (ssb *shardStorageBootstrapper) cleanupNotarizedStorageForHigherNoncesIfExist(
+	crossNotarizedHeaders []bootstrapStorage.BootstrapHeaderInfo,
+) {
+	var numConsecutiveNoncesNotFound int
+
+	nonce, err := getCrossNotarizedHeaderNonce(crossNotarizedHeaders)
+	if err != nil {
+		log.Warn("cleanupNotarizedStorageForHigherNoncesIfExist", "error", err.Error())
+		return
+	}
+
+	log.Debug("cleanup notarized storage higher than nonce", "nonce", nonce)
+
+	for {
+		nonce++
+
+		metaBlock, metaBlockHash, err := process.GetMetaHeaderFromStorageWithNonce(
+			nonce,
+			ssb.store,
+			ssb.uint64Converter,
+			ssb.marshalizer,
+		)
+		if err != nil {
+			log.Debug("meta block is not found in MetaHdrNonceHashDataUnit storage",
+				"nonce", nonce)
+
+			numConsecutiveNoncesNotFound++
+			if numConsecutiveNoncesNotFound > maxNumOfConsecutiveNoncesNotFoundAccepted {
+				break
+			}
+
+			continue
+		}
+
+		numConsecutiveNoncesNotFound = 0
+		nonceToByteSlice := ssb.uint64Converter.ToByteSlice(metaBlock.GetNonce())
+		err = ssb.store.GetStorer(dataRetriever.MetaHdrNonceHashDataUnit).Remove(nonceToByteSlice)
+		if err != nil {
+			log.Debug("meta block was not removed from MetaHdrNonceHashDataUnit storage",
+				"shardId", metaBlock.GetShardID(),
+				"nonce", metaBlock.GetNonce(),
+				"hash", metaBlockHash,
+				"error", err.Error())
+		}
+
+		err = ssb.store.GetStorer(dataRetriever.MetaBlockUnit).Remove(metaBlockHash)
+		if err != nil {
+			log.Debug("meta block was not removed from MetaBlockUnit storage",
+				"shardId", metaBlock.GetShardID(),
+				"nonce", metaBlock.GetNonce(),
+				"hash", metaBlockHash,
+				"error", err.Error())
+		}
+	}
+}
+
+func getCrossNotarizedHeaderNonce(crossNotarizedHeaders []bootstrapStorage.BootstrapHeaderInfo) (uint64, error) {
+	for _, crossNotarizedHeader := range crossNotarizedHeaders {
+		if crossNotarizedHeader.ShardId != core.MetachainShardId {
+			continue
+		}
+
+		log.Debug("cross notarized header",
+			"shard", crossNotarizedHeader.ShardId,
+			"epoch", crossNotarizedHeader.Epoch,
+			"nonce", crossNotarizedHeader.Nonce,
+			"hash", crossNotarizedHeader.Hash)
+
+		return crossNotarizedHeader.Nonce, nil
```

Comment:
do we need to return only the higher nonce from the provided metablock?
In other words: does the crossNotarizedHeaders param can contain 2 or more metablock?

Reply:
No, crossNotarizedHeaders is a list which contains only one header for each shard. It is the last finalized and committed cross notarized block.